internal/pkg/deploy/cloudformation/stack/validate.go

```diff
@@ -100,6 +114,154 @@ func validateSidecarMountPoints(in []manifest.SidecarMountPoint) error {
 	return nil
 }
 
+func validateSidecarDependsOn(in manifest.SidecarConfig, sidecarName string, s convertSidecarOpts) error {
+	if in.DependsOn == nil {
+		return nil
+	}
+
+	for name, status := range in.DependsOn {
+		// status must be one of < start | complete | success >
+		if !isValidStatus(status) {
+			return errInvalidDependsOnStatus
+		}
+		// essential containers must have 'start' as a status
+		if name == s.workloadName || s.sidecarConfig[name].Essential == nil || aws.BoolValue(s.sidecarConfig[name].Essential) {
+			if status != "start" {
+				return errEssentialContainerStatus
+			}
+		}
+	}
+
+	return nil
+}
+
+func validateNoCircularDependencies(s convertSidecarOpts) error {
+	used := make(map[string]bool)
+	path := make(map[string]bool)
+	dependencies, err := buildDependencyGraph(s)
+
+	// don't let nonexistent containers pass
+	if err != nil {
+		return err
+	}
+
+	// don't let circular dependencies happen
+	for node := range dependencies.nodes {
+		if !used[node] && hasCycles(dependencies, used, path, node) {
+			return errCircularDependency
+		}
+	}
+
+	return nil
+}
+
+func hasCycles(graph *Graph, used map[string]bool, path map[string]bool, currNode string) bool {
+	used[currNode] = true
+	path[currNode] = true
+
+	for _, node := range graph.nodes[currNode] {
+		if !used[node] && hasCycles(graph, used, path, node) {
+			return true
+		} else if path[node] {
+			return true
+		}
+	}
+
+	path[currNode] = false
+	return false
+}
+
+func buildDependencyGraph(s convertSidecarOpts) (*Graph, error) {
+	dependencyGraph := Graph{nodes: make(map[string][]string)}
+
+	// sidecar dependencies
+	for name, sidecar := range s.sidecarConfig {
+		// add any existing dependency to the graph
+		if len(sidecar.DependsOn) != 0 {
+			for dep := range sidecar.DependsOn {
+				// containers being depended on must exist
+				if s.sidecarConfig[dep] == nil && dep != s.workloadName {
+					return nil, errInvalidContainer
+				}
+
+				dependencyGraph.Add(name, dep)
+			}
+		}
+	}
+
+	// add any image dependencies to the graph
+	if len(s.imageConfig.DependsOn) != 0 {
+		for dep := range s.imageConfig.DependsOn {
+			// containers being depended on must exist
+			if s.sidecarConfig[dep] == nil && dep != s.workloadName {
+				return nil, errInvalidContainer
+			}
+
+			dependencyGraph.Add(s.workloadName, dep)
+		}
+	}
+
+	return &dependencyGraph, nil
+}
+
+type Graph struct {
```

Comment:
nit this struct and its methods don't need to be public.

Reply:
Now not public.